### ISSUE TITLE
Feature Request: Make it possible for `transform` to return issues

### ISSUE THREAD

**Issue by duailibe** [fabian-hiller/valibot#182]
When it comes to custom validation or transformation, AFAICT, Valibot's  `transform()`, `custom()` and `special()` can either validate or transform something but not both.

When using zod, one pattern that I've been using sometimes is to use `transform()` right after a validation and perform custom validation and transformation.

One example is being able to parse and validate one JSON string (code below taken from [a little library I wrote](https://github.com/duailibe/zod-extras/blob/main/src/jsonstring.js)):

```js
export function zjsonstring(schema) {
  const validateJson = z.string().transform((val, ctx) => {
    try {
      const parsed = JSON.parse(val);
      return parsed;
    } catch (e) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: e instanceof Error ? e.message : "Invalid JSON"
      });
      return z.NEVER;
    }
  });
```

So I can just use `zjsonstring({ foo: z.string(), bar: z.number() })` more practically. But there are other cases in the wild, like Kent C Dodds use `transform()` to [perform a login *and* return the session](https://github.com/epicweb-dev/epic-stack/blob/af13de31239ba4e2f13eca894ac4698dea52c48b/app/routes/_auth%2B/login.tsx#L200-L213).

Given current valibot architecture, this would be a breaking change as the new API would need to return output or issues, and not output directly like today.

**Comment by duailibe** [comment]
Given valibot's architecture, not sure if this would require a new API since the passed `action` would need to return either `{ issue }` or `{ output }` and the direct output like today.

**Comment by fabian-hiller** [comment]
Thank you for creating this issue. Basically you need the ability to add or return an issue in our `transform` method. You are right that if we stick to the current approach in our pipeline, we would need an object with `output` or `issues`, which would be a breaking change. Alternatively we could consider adding the ability to throw a `ValiError` in `transform`.

**Comment by duailibe** [comment]
Thanks! I can take a shot at a PR today (actually already done it in a fork that I'm using and I can try to contribute it upstream).

Do you have any preferences between a breaking change vs. throwing a ValiError ?

**Comment by fabian-hiller** [comment]
Currently I think it is a better API to just be able to return the transformed data without `return { outout: ... }`. In the pipeline it is Ok, because here mostly the validation and transformation functions are used. What do you think about this?

**Comment by duailibe** [comment]
Throwing a `ValiError` in `transform` seems like a good alternative and easy to add, I just don't like very much the lack of parity between pipeline and transformations (but maybe that's just me that needs some getting used to - they are already different in many senses)

**Comment by duailibe** [comment]
Also maybe I wouldn't necessarily throw a ValiError since the transform would still need to catch to "merge" the issue with the info object, so I guess a `throw getTransformIssue(transform, error, input?)`.

I get this isn't a little unorthodox (I personally use sometimes 😅) so I understand if you're not on board, just throwing it (pun not intended 😅) out there.

**Comment by fabian-hiller** [comment]
Another solution might be to validate your JSON in our pipeline (maybe with a new `json` validation function) and then transform it in `transform`. Since `transform` is only executed if there are no issues, this should work.

**Comment by duailibe** [comment]
@fabian-hiller the problem with that approach is running the same code twice (once for validation, other for transformation). For the JSON case that would be fine (and a new type might even make more sense - but could also be done in userland).

For more complex cases (like the authentication I case wrote above), running the check twice wouldn't work.

**Comment by fabian-hiller** [comment]
Mhm... that's probably true. At the moment I don't know exactly how I would proceed. Every solution seems to be not optimal.

**Comment by fabian-hiller** [comment]
Perhaps a similar approach with a function passed as an argument as in Zod is a better compromise.

**Comment by duailibe** [comment]
Do you mean, instead of:

```
.transform(value => newValue);
```

you're suggesting:

```
.transform((value, addIssue) => {
  if (something) {
    addIssue(issue);
    return
  }
  return newValue;
});
```

?

**Comment by fabian-hiller** [comment]
Yes for example.

**Comment by cbodin** [comment]
I'm in the process of validating search parameters and my first assumption was that the transform function would actually accept an error message and handle thrown errors by default.

This is especially useful for Temporal.PlainTime/PlainDate`.from()` as they throw if the argument cannot be parsed correctly.

Throwing a `ValiError` would be really useful but also providing a default error message for other errors would awesome.
My current schema looks like this:
```ts
export const PlainDateStringSchema = transform(
  string([
    custom((input) => {
      try {
        PlainDate.from(input);
        return true;
      } catch (_) {
        return false;
      }
    }, "Must be a valid date."),
  ]),
  (input) => PlainDate.from(input),
);
```

But could be reduced to something like this if errors were handled automatically:
```ts
transform(
  string(),
  (input) => PlainDate.from(input),
  "Must be a valid date."
)
```

**Comment by fabian-hiller** [comment]
Thank you very much! I am currently rewriting Valibot based on discussion #463. I will take a look at your code once that is done. Feel free to ping me if I forget to answer.

**Comment by maranomynet** [comment]
I would very much like to be able to either throw a `ValiError` from inside a `transform` action, or have a zod-like `addIssue` method and be able to return a `v.NEVER` symbol.

I have a transform function which returns a branded string value or undefined (if the transformation fails) and I need to be able to error on the undefined case.

```ts
declare const _BrandedFooString__Brand: unique symbol;
type BrandedFooString = string & { [_BrandedFooString__Brand]: true };

declare function parseFoo(str: string): BrandedFooString | undefined;

const fooSchema = v.transform(v.string(), parseFoo);
//      ^? v.SchemaWithTransform<v.StringSchema<string>, BrandedFooString | undefined>
```

If I try to further validate the transformed value, to get rid of the `undefined` value, I lose the braned type, and the type still includes `undefined`

```ts
const fooSchema = v.transform(v.string(), parseFoo, v.string('Not valid foo'));
//      ^? v.SchemaWithTransform<v.StringSchema<string>, string | undefined>

```

The Type I want is `v.SchemaWithTransform<v.StringSchema<string>, BrandedFooString>`

The solution I'm left with at the moment is this borderline abomination:

```ts
const fooSchema = v.transform(
  v.transform(
    v.string(),
    parseFoo,
    v.string('Not valid foo')
  ),
  (v) => v as BrandedFooString  // ack, manual type assertion
);
```

What I'd love to be able to do is this:

```ts
const fooSchema = v.transform(
  v.string(),
  (value) => {
    const foo = parseFoo(value);
    if (foo === undefined) {
      throw new v.ValiTransformError('Not valid foo');
    }
    return foo;
  }
);
```

**Comment by fabian-hiller** [comment]
For now, I recommend using the `pipe` of `string` to validate, e.g. with `regex`, that `parseFoo` never returns `undefined`. But I will investigate the possibility of adding issues in `transform` after we merge the ongoing rewrite of the entire library in #502.

**Comment by maranomynet** [comment]
ok, if `parseFoo` performs minor cleanup/transformation on the input before branding and returning it
...how would you use the pipe for that?

Is there a page in the docs I might have missed?

**Comment by maranomynet** [comment]
Sorry, it's before my coffe o'clock here. I understand what you mean now.  
Essentially this, right?

```ts
const fooSchema = v.transform(
  v.transform(
    v.string(),
    parseFoo,
    [v.regex(/./)]
  ),
  (v) => v as BrandedFooString  // ack, manual type assertion
);
```

**Comment by fabian-hiller** [comment]
Right now you have two options. The first is to validate the input before executing your `transform` function. The second is to validate the output of the transformation.

```ts
// Option 1
const Schema = v.transform(v.string([v.decimal()]), parseInt);

// Option 2
const Schema = v.transform(v.string(), parseInt, v.number());
```

**Comment by jamiebuilds** [comment]
I would like it if thrown errors are not automatically handled. I would like to be able to distinguish between issues I'm intentionally returning and errors that are problems in my schema code.

That also makes it a bit more natural to pass a message.

```ts
function toNumber(message = "Invalid number syntax") {
  return v.transform(input => {
    let trimmed = input.trim()
    if (trimmed === "") {
      return null
    }
    let parsed = Number(trimmed)
    if (!Number.isFinite(parsed)) {
      return v.issue(message)
    }
    return parsed
  })
}

let urlIntParam = v.nullish(
  v.pipe(
    v.string()
    toNumber("custom message"),
    v.nullable(v.pipe(v.number(), v.safeInteger()))
  ),
  null,
)
```

You may want to consider the experience adding types for these actions/messages though:

```ts
// v.action() would handle the types all internally for you
const toNumber = v.action((message = "Invalid number syntax") => {
  return v.transform(input => {
    // ...
    return v.issue(message)
  })
})

toNumber("custom message")
```

**Comment by fabian-hiller** [comment]
@jamiebuilds this is related to #597

**Comment by fabian-hiller** [comment]
This is now possible with our new `rawTransform` API.
